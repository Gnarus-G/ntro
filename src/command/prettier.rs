use anyhow::Context;
use std::{path::Path, process};

use which::which;

use super::pm::PackageManager;

pub fn prettier(file_extension: &str) -> anyhow::Result<process::Child> {
    let pm = PackageManager::from_current_project();

    // pretend file name, doesn't need to exist,
    // but its extension describe how prettier
    // should work
    let file_name = &Path::new("temp").with_extension(file_extension);

    let package_manager_executable = pm
        .map(PackageManager::into_executor_name)
        .context("no package manager detected, neither pnpm, npm, nor yarn")
        .map(|exe| (exe, vec!["prettier", "--stdin-filepath"]));

    let (exe, args) = which("prettierd")
        .map(|_| ("prettierd", vec![]))
        .or(package_manager_executable.context("failed to find prettierd in the path, and couldn't use a package manager to execute prettier with"))?;

    let child = process::Command::new(exe)
        .args(&args)
        .arg(file_name)
        .stdin(process::Stdio::piped())
        .stdout(process::Stdio::piped())
        .spawn()
        .with_context(|| {
            format!(
                "failed to spawn prettier with command: {} {}",
                exe,
                [args, vec![&file_name.to_string_lossy()]]
                    .concat()
                    .join(" ")
            )
        })?;

    Ok(child)
}
