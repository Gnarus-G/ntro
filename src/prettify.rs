use anyhow::{anyhow, Context};
use ntro::pm::PackageManager;
use std::{io::Write, path::Path, process};

use which::which;

pub fn prettify(file: &[u8], file_name: &Path) -> anyhow::Result<Vec<u8>> {
    let mut prettier = prettier(file_name)?;

    let mut prettier_stdin = prettier.stdin.take().ok_or(anyhow!(
        "failed to open stdin to pass file contents to prettier"
    ))?;

    prettier_stdin.write_all(file)?;

    // Finish (close file handle)
    drop(prettier_stdin);

    let output = prettier.wait_with_output()?;

    if output.status.success() {
        Ok(output.stdout)
    } else {
        Err(anyhow!(
            "{}",
            String::from_utf8_lossy(&[output.stdout, output.stderr].concat())
        )
        .context("exited prettier execution with a fail status"))
    }
}

fn prettier(file_name: &Path) -> anyhow::Result<process::Child> {
    let pm = PackageManager::from_current_project();

    let package_manager_executable = pm
        .map(PackageManager::into_executor_name)
        .context("no package manager detected, neither pnpm, npm, nor yarn")
        .map(|exe| (exe, vec!["prettier", "--stdin-filepath"]));

    let (exe, args) = which("prettierd")
        .map(|_| ("prettierd", vec![]))
        .or(package_manager_executable.context("failed to find prettierd in the path, and couldn't use a package manager to execute prettier with"))?;

    let child = process::Command::new(exe)
        .args(&args)
        .arg(file_name)
        .stdin(process::Stdio::piped())
        .stdout(process::Stdio::piped())
        .spawn()
        .with_context(|| {
            format!(
                "failed to spawn prettier with command: {} {}",
                exe,
                [args, vec![&file_name.to_string_lossy()]]
                    .concat()
                    .join(" ")
            )
        })?;

    Ok(child)
}
