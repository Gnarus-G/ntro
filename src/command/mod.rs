mod pm;
mod prettier;

use std::{
    fs::File,
    io::{BufReader, Write},
    process::Command,
};

use anyhow::{anyhow, Context, Result};
use serde_json::Value;

use crate::command::pm::PackageManager;

use self::prettier::prettier;

pub fn prettify<E: AsRef<str>>(content: &[u8], file_extension: E) -> anyhow::Result<Vec<u8>> {
    let mut prettier = prettier(file_extension.as_ref())?;

    let mut prettier_stdin = prettier.stdin.take().ok_or(anyhow!(
        "failed to open stdin to pass file contents to prettier"
    ))?;

    prettier_stdin.write_all(content)?;

    // Finish (close file handle)
    drop(prettier_stdin);

    let output = prettier.wait_with_output()?;

    if output.status.success() {
        Ok(output.stdout)
    } else {
        Err(anyhow!(
            "{}",
            String::from_utf8_lossy(&[output.stdout, output.stderr].concat())
        )
        .context("exited prettier execution with a fail status"))
    }
}

pub fn npm_install(package: &str) -> Result<()> {
    let package_info: Value = File::open("./package.json")
        .context("couldn't open package.json")
        .map(BufReader::new)
        .and_then(|reader| serde_json::from_reader(reader).context("failed to parse package.json"))
        .context("failed to read package.json")?;

    if package_info
        .get("dependencies")
        .and_then(|deps| deps.get(package))
        .is_some()
    {
        return Ok(());
    }

    log::info!("installing {}...", package);
    let out = PackageManager::from_current_project()
        .ok_or(anyhow!("couldn't get package manager from current project"))
        .or(PackageManager::from_global())
        .map(|pm| match pm {
            PackageManager::Pnpm => ("pnpm", "add"),
            PackageManager::Yarn => ("yarn", "add"),
            PackageManager::Npm => ("npm", "i"),
        })
        .and_then(|(exe, arg)| {
            Command::new(exe)
                .arg(arg)
                .arg(package)
                .output()
                .with_context(|| {
                    format!("failed to execute installation with package manager: {exe}")
                })
        })?;

    if !out.status.success() {
        return Err(anyhow!(
            "installation failed with exit code {:?}",
            out.status.code()
        ));
    }

    Ok(())
}
