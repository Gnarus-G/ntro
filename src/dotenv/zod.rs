use serde_json::Value;
use std::{
    collections::BTreeMap,
    fmt::Display,
    fs::File,
    io::{BufReader, BufWriter, Read, Write},
    path::{Path, PathBuf},
};

use anyhow::{anyhow, Context, Result};

use crate::command::prettify;

use super::{
    parse::{parse_variables_with_type_hints, Variable},
    typehint_parser::TypeHint,
};

use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("found {a} is different from {b}")]
    ConflictingTypes { a: TypeHintAt, b: TypeHintAt },
}

#[derive(Debug)]
pub struct TypeHintAt {
    pub th: TypeHint,
    pub file: PathBuf,
    pub line: usize,
}

impl Display for TypeHintAt {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "type {:?} in {}:{}",
            self.th,
            self.file.to_string_lossy(),
            self.line,
        )
    }
}

impl From<(&PathBuf, &(TypeHint, usize))> for TypeHintAt {
    fn from(value: (&PathBuf, &(TypeHint, usize))) -> Self {
        Self {
            th: value.1 .0.clone(),
            file: value.0.clone(),
            line: value.1 .1,
        }
    }
}

pub fn generate_zod_schema(files: &[PathBuf]) -> Result<String> {
    let mut map: BTreeMap<String, (Variable, PathBuf)> = BTreeMap::new();

    let parse = |text: String, file_name| {
        return parse_variables_with_type_hints(&text)
            .into_iter()
            .map(|var| (var, file_name))
            .collect::<Vec<_>>();
    };

    let variables = files
        .iter()
        .map(|file| {
            File::open(file)
                .map(BufReader::new)
                .and_then(|mut rdr| {
                    let mut buf = String::new();
                    rdr.read_to_string(&mut buf).map(|_| buf)
                })
                .context(format!("failed read {file:?}"))
                .map(|text| parse(text, file))
        })
        .filter_map(|result| {
            if let Err(e) = &result {
                eprintln!("{e:?}");
            }
            result.ok()
        })
        .flatten();

    for (var, file_name) in variables {
        if let Some((v, ofile_name)) = map.get(&var.key) {
            if let (Some(lt), Some(rt)) = (&v.type_hint, &var.type_hint) {
                if lt != rt {
                    return Err(ParseError::ConflictingTypes {
                        a: (ofile_name, lt).into(),
                        b: (file_name, rt).into(),
                    })
                    .context(
                        "found some conflicting types while parsing variables with type hints",
                    );
                }
            }
        }

        map.insert(var.key.clone(), (var, file_name.clone()));
    }

    let vars = map.into_values().map(|value| value.0).collect::<Vec<_>>();

    let next_public_vars = vars.iter().filter(|&v| v.is_public()).collect::<Vec<_>>();
    let other_vars = vars.iter().filter(|v| !v.is_public()).collect::<Vec<_>>();

    let to_field_schema = |var: &&Variable| -> String {
        format!(
            r#"    {}: {},"#,
            var.key,
            match &var.type_hint {
                Some(th) => match &th.0 {
                    super::typehint_parser::TypeHint::String => "z.coerce.string()".to_string(),
                    super::typehint_parser::TypeHint::Number => "z.coerce.number()".to_string(),
                    super::typehint_parser::TypeHint::Boolean => "z.coerce.boolean()".to_string(),
                    super::typehint_parser::TypeHint::Union(values) =>
                        format!("z.enum([{}])", values.join(",")),
                },
                None => "z.string()".to_string(),
            }
        )
    };

    let js_code = include_str!("module.ts");

    let js_import_line: &str = js_code
        .lines()
        .next()
        .expect("should have an import line at the top of the js implementation");

    let js_impl = js_code
        .lines()
        .skip_while(|line| !line.contains("/* --- MAIN IMPLEMENTATION BELOW --- */"))
        .skip(1)
        .collect::<Vec<_>>()
        .join("\n");

    let output = format!(
        r#"
{js_import_line}

const clientEnvSchemas = {{
{}
}}

const serverEnvSchemas = {{
    ...clientEnvSchemas,
{}
}}

{js_impl}

const processEnv = {{
{}
}}
               "#,
        next_public_vars
            .iter()
            .map(to_field_schema)
            .collect::<Vec<_>>()
            .join("\n"),
        other_vars
            .iter()
            .map(to_field_schema)
            .collect::<Vec<_>>()
            .join("\n"),
        vars.iter()
            .map(|var| format!("   {}: process.env.{},", var.key, var.key))
            .collect::<Vec<_>>()
            .join("\n"),
    );

    Ok(output)
}

pub fn add_tsconfig_path<P: AsRef<Path>>(path: P) -> Result<()> {
    let mut ts_config: Value = File::open("./tsconfig.json")
        .context("couldn't open tsconfig.json")
        .map(BufReader::new)
        .and_then(|reader| serde_json::from_reader(reader).context("failed to parse tsconfig.json"))
        .context("failed to read tsconfig.json")?;

    ts_config
        .get_mut("compilerOptions")
        .context("couldn't find compilerOptions in tsconfig.json")
        .and_then(|paths| {
            paths
                .get_mut("paths")
                .and_then(|node| node.as_object_mut())
                .map(|paths| {
                    paths.insert(
                        "$env".to_string(),
                        Value::Array(vec![Value::String(
                            path.as_ref().to_string_lossy().to_string(),
                        )]),
                    )
                })
                .ok_or(anyhow!("failed to add $env as a path on tsconfig.json"))
        })?;

    File::create("./tsconfig.json")
        .context("failed to open tsconfig.json")
        .map(BufWriter::new)
        .and_then(|mut w| {
            let new_content = serde_json::to_string::<Value>(&ts_config)?;
            let pretty = prettify(new_content.as_bytes(), ".json")?;
            w.write_all(&pretty)
                .context("failed to write to tsconfig.json")?;
            w.flush()
                .context("failed to flush updated tsconfig.json contents")
        })
        .context("failed to update tsconfig.json contents")?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::{assert_debug_snapshot, assert_display_snapshot};

    use crate::dotenv::zod::generate_zod_schema;

    #[test]
    fn zod_schema_gen() {
        let output = generate_zod_schema(&[
            PathBuf::from("src/dotenv/.env.test"),
            PathBuf::from("src/dotenv/.env.test2"),
        ])
        .unwrap();
        assert_display_snapshot!(output);
    }

    #[test]
    fn zod_schema_gen_with_type_conflicts() {
        let output = generate_zod_schema(&[
            PathBuf::from("src/dotenv/.env.test"),
            PathBuf::from("src/dotenv/.env.dupes.test"),
        ])
        .unwrap_err();

        assert_debug_snapshot!(output);
    }
}
