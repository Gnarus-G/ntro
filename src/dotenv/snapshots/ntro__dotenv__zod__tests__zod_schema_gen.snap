---
source: src/dotenv/zod.rs
expression: output
---

import z, { ZodTypeAny } from "zod";

const clientEnvSchemas = {

}

const serverEnvSchemas = {
    ...clientEnvSchemas,
    KEY: z.string(),
    KEY_Value: z.coerce.number(),
    NAME: z.string(),
    NAME2: z.enum(['a','b']),
    NAME5: z.string(),
    keys: z.string(),
    keys2: z.coerce.string(),
    keys2Da: z.enum(['city','townhall']),
}


export const clientEnv: z.infer<z.ZodObject<typeof clientEnvSchemas>> =
  new Proxy({} as any, {
    get(_, prop: string) {
      return lookupEnv(prop, clientEnvSchemas, () => {
        throw new Error(
          `${prop} is not defined for client side environment variables.`
        );
      });
    },
  });

export const env: z.infer<z.ZodObject<typeof serverEnvSchemas>> = new Proxy(
  {} as any,
  {
    get(_, prop: string) {
      if (prop.startsWith("NEXT_PUBLIC_")) {
        return Reflect.get(clientEnv, prop);
      }
      return lookupEnv(prop, serverEnvSchemas, () => {
        throw new Error(
          `${prop} is not defined for server side environment variables.`
        );
      });
    },
  }
);

const cache: Record<string, unknown> = {};

function lookupEnv<T extends Record<string, ZodTypeAny>>(
  prop: string,
  parsers: T,
  onNotFound: () => never
) {
  if (prop in cache) {
    return cache[prop];
  }

  try {
    if (prop in parsers) {
      const parsed = parsers[prop as keyof typeof parsers].parse(
        processEnv[prop as keyof typeof processEnv],
        { path: [prop] }
      );

      cache[prop] = parsed;

      return parsed;
    }
    onNotFound();
  } catch (e) {
    throw new BadEnvError(`failed to read ${prop} from proccess.env`, e);
  }
}

class BadEnvError extends Error {
  constructor(public message: string, public cause: unknown) {
    super(message);
    if (cause instanceof Error) {
      this.message = [message, cause].join("\n â†³ ");
    }
  }
}

const processEnv = {
   KEY: process.env.KEY,
   KEY_Value: process.env.KEY_Value,
   NAME: process.env.NAME,
   NAME2: process.env.NAME2,
   NAME5: process.env.NAME5,
   keys: process.env.keys,
   keys2: process.env.keys2,
   keys2Da: process.env.keys2Da,
}
               
